File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        int index = 0; 
        ListNode currNode = head; 
        HashMap<ListNode, Integer> nodeMap = new HashMap<>(); 

        while(currNode != null){ 
            if(!nodeMap.containsKey(currNode)){ 
                nodeMap.put(currNode, index); 
            } else { 
                return true; 
            }

            currNode = currNode.next; 
            index++; 

        }
        return false; 
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        dummy = Node()
        curr = head

        hashMap = {} 

        while(curr.next): 
            hashMap[curr.val] = ListNode(curr.val, None, None)
            curr = curr.next 

        curr = head 
        tail = dummy 

        while(curr.next): 
            nodeToPut = hashMap[curr.val]
            nodeToPut.next = hashMap[curr.next.val] if curr.next else None 
            nodeToPut.random = hashMap[curr.random.val] if curr.random else None

        return dummy.next


